lib/puppet/type/zfs_share.rb

```diff
@@ -0,0 +1,234 @@
+Puppet::Type.newtype(:zfs_share) do
+  @doc = "Manage zfs shares"
+
+  require 'puppet/parameter/boolean'
+
+  ensurable
+
+  # zfs share documentation
+  # https://docs.oracle.com/cd/E26502_01/html/E29031/zfs-share-1m.html
+
+  newparam(:name) do
+    desc "Name of zfs file system"
+    isnamevar
+    validate do |value|
+      if value =~ /^\//
+        raise ArgumentError , "#{value} is not a valid zfs title"
+      end
+    end
+  end
+
+  newparam(:share_title) do
+    desc "Creates a named share with the share title"
+    validate do |value|
+      if value =~ /\s/
+        raise ArgumentError , "#{value} is not a valid share title"
+      end
+    end
+  end
+
+  newproperty(:nfs) do
+    desc "Enable or disable nfs"
+    defaultto 'on'
+    newvalues(:on, :off)
+  end
+
+  newproperty(:nfs_log) do
+    desc "Enables NFSv2 or NFSv3 server logging for the specified file system. The tag is defined in the /etc/nfs/nfslog.conf file. If no tag is specified, the default values associated with the global tag in the /etc/nfs/nfslog.conf file is used."
+  end
+
+  newproperty(:nosub) do
+    desc "Prevents NFSv2 or NFSv3 clients from mounting subdirectories of shared directories."
+    newvalues(:on, :off)
+  end
+
+  newparam(:guestok) do
+    desc "Whether to allow guest access to smb shares"
+    newvalues(:on, :off)
+  end
+
+  newproperty(:sec_none_rw, :array_matching => :all) do
+    validate do |value|
+      unless value =~ /^\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\/\d{1,2}$|\*/
+        raise ArgumentError , "#{value} is not a valid IP address."
```

Comment:
Do we use it in that way anywhere to your knowledge? It may be worth mentioning that I'm not necessarily aiming to support every possible use case, just the ones we use. If others would like to expand upon this, or we need to, I certainly welcome that. I'm trying to identify a balance between fully-featured and the features that we need/use.

If you would like to see this expanded to support those use cases then it is possible. Doing so will require some effort to correctly manipulate those entries, and since I do not use them in that way I have not been incentivized to devise test cases for that usage.